### PR TITLE
Update Bookmarks UI and extract Bookmarks database

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/72649045549333/1202533376181935/f
Tech Design URL:
CC:

**Description**:

**Steps to test this PR**:
1.  Test Migration.  Using a build from develop create bookmarks, favorites, folders.  Install new version and check they are migrated successfully (favorite merged with bookmark if same url, old database cleaned etc).
2. Test Clean install. Using a clean build from develop, ensure app starts and you can create bookmarks, favorites, folders, etc. 
3. Check that changes are preserved across app restarts.
4. Check dependency chain and how Main view is now instantiated. Validate app start/restart state and lock/auto clear.
5. Check add/edit bookmark / favorite from browsing menu and menu updates to show correct state on change.
6. Check bookmark manager view shows bookmarks and folder navigation works.
7. Check bookmark manager view allows editing of bookmarks and folders.
8. Use swipe actions in bookmark manager view to set as favorite and delete.
9. Search functionality in bookmark manager shows expected bookmarks and swipe to favorite/delete works. Search updates when DB state changes.
10. When using address bar bookmarks top 2 matching bookmarks appear in the suggestions, and what you see matches DB latest state.
11. Favorites can be re-ordered on the favorites tab.
12. Switching between bookmarks and favorites updates the UI correctly.
13. Add all bookmarks from from tab switcher.
14. Favicons are saved as expected for bookmarks when they are added
15. Favorites are show in overlay (not re-orderable or editable)
16. Favorites show on home screen and are re-orderable 
17. Favorites show edit / remove menu when long pressed
18. Bookmarks import & export works as expected.
 

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: DuckDuckGo/Base.lproj/Bookmarks.storyboard [hunk 1]
```diff
@@ -16,11 +17,56 @@
...
+                                <subviews>
+                                    <imageView clipsSubviews="YES" userInteractionEnabled="NO" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" image="BookmarkFavoritesHero" translatesAutoresizingMaskIntoConstraints="NO" id="FRF-MC-5eG">
+                                        <rect key="frame" x="0.0" y="0.0" width="331" height="115.5"/>
+                                        <constraints>
+                                            <constraint firstAttribute="height" constant="115.40000000000001" id="nKK-XC-tDl"/>
+                                        </constraints>
+                                    </imageView>
+                                    <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="No favorites added yet" textAlignment="center" lineBreakMode="tailTruncation" numberOfLines="2" baselineAdjustment="alignBaselines" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="y8y-oO-xBm">
```
Comment (samsymons) [inline DuckDuckGo/Base.lproj/Bookmarks.storyboard:42]: I don't think this is using the localized string – I see we do have `UserText.emptyFavorites`, but it isn't used and likely should be here

Comment (brindy) [reply DuckDuckGo/Base.lproj/Bookmarks.storyboard:42]: Just updated translations actually.  I've removed this.  Thanks!

File: DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift [hunk 2]
```diff
@@ -236,37 +251,30 @@ extension TabViewController {
...
-        bookmarksManager.saveNewFavorite(withTitle: link.title ?? "", url: link.url) { _, _ in
-            DispatchQueue.main.async {
-                ActionMessageView.present(message: UserText.webSaveFavoriteDone, actionTitle: UserText.actionGenericUndo, onAction: {
-                    self.performRemoveFavoriteAction(for: link)
-                })
-            }
-        }
+    private func performAddFavoriteAction(for link: Link,
+                                          with bookmarksInterface: MenuBookmarksInteracting) {
+        bookmarksInterface.createOrToggleFavorite(title: link.title ?? "", url: link.url)
```
Comment (samsymons) [inline DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift:263]: It's possible for this to be out of sync when you're loading a new page - I was able to get one bookmark with a blank title, and another with the title of the previous page. However, this isn't a regression introduced in this PR, and it takes fairly precise timing, I don't consider it a blocker.

Comment (brindy) [reply DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift:263]: Will create a separate task for this, thanks.

Comment (brindy) [reply DuckDuckGo/TabViewControllerBrowsingMenuExtension.swift:263]: https://app.asana.com/0/414709148257752/1203462254653688/f

File: Core/Link.swift [hunk 3]
```diff
@@ -37,17 +37,17 @@ public class Link: NSObject, NSCoding {
     public let url: URL
     public let localFileURL: URL?
     
-    public var displayTitle: String? {
+    public var displayTitle: String {
```
Comment (samsymons) [inline Core/Link.swift:40]: I see 4 new warnings in the codebase now that this isn't optional - they were defaulting to the empty string, but don't need to do that any more.

* TabViewGridCell
* TabViewListCell

File: DuckDuckGo/HomeViewController.swift [hunk 4]
```diff
@@ -19,7 +19,9 @@
 
 import UIKit
 import Core
+import Bookmarks
 import os.log
+import Bookmarks
```
Comment (samsymons) [inline DuckDuckGo/HomeViewController.swift:23]: Duplicate import here

**Review by samsymons — APPROVED** [review]
This is looking good. I like the way the new bookmarks library is structured 👏 

I've tested every feature related to bookmarks and it's all working well.

**Review by samsymons — APPROVED** [review]
The new changes seem okay, though we have some SwiftLint and unit test failures at the moment.

**Pull request merged**